### ISSUE TITLE
BUG: Deleted sprite not actually deleted, reappears when file reloaded.

### ISSUE THREAD

**Issue by deader69** [jmoenig/Snap--Build-Your-Own-Blocks#795]
http://snap.berkeley.edu/snapsource/snap.html#present:Username=deader-tcssm&ProjectName=MorseCode

I've noticed that the Sprite named 'when' doesn't actually get deleted in my project.
To repro the bug:
1) Load the project, 
2) Right-click the 'when' Sprite and delete it.
3) Save (or Export) the file.
4) Load (or Import) the file.
5) Notice that the Sprite named 'when' is still there.

I'm going to try other Sprites/Projects later (very slow wifi right now).

Note that I'm using/abusing the [__ of __] block, to use Sprites as classes/modules, so maybe I have a dangling reference to the 'when' sprite. (Although I think [__of__] just holds a string, isn't 'known' by Snap!).

**Comment by jmoenig** [comment]
Wow, this is another of those things you've managed to pull off that make me wonder whether you've used experimental or hidden features such as dev mode or editing the XML file  with a text editor, or exploiting a funky glitch such as abusing the block editor as a clipboard for sprite-local blocks when switching to another one.

Actually I presume it's the latter.

How did you manage to put the "check edge" block, which is local to the "when" sprite, into the block definition of "plot end of word" of the "plot" sprite? The only way this seems possible is by exploiting the glitch above.

I guess I should force all block editors on local block definitions to close whenever the user switches among sprites... This is somewhat of a bummer, because it's often convenient to leave them open while looking up something in another sprite, but - as you've shown - it can also lead to spaghetti references among sprites :-(

**Comment by deader69** [comment]
Yes, I've been "..abusing the block editor as a clipboard for sprite-local blocks when switching to another one."
Guilty as charged! 

And/or just copying a Sprite before starting a major re-architecting of it, wanting to keep the old thing working whilst developing the new.

And yes, there's some ugly/experimental stuff in my code, trying to figure out the best way to organize things. I'm trying to use Sprites as modules, so that I can hide 90% of the low-level helper routines, and expose only the high-level API.

The problem is that global block defs cannot contain references to local defs... 

Anyway, you can see there are a whole lot of transformations required to convert a text string to a Morse code sequence. If I made ALL those routines global, it'd overwhelm the average coder (and really clutter up the Snap! layout with a long list of custom block defs).

So I make them all private, then give global access via 'function pointer'; so that other modules can call them via [call 'translate word' of 'translate'].

I really just want to provide users with a couple of high level 'translate' routines, without exposing the numerous subroutines (index of upper case unicode, index of lower case unicode, etc ad nauseum).

Is there any way to do that cleanly?

**Comment by deader69** [comment]
BTW, I don't expect you to change Snap! to fix this... just thought it'd be of interest, might be revealing some important underlying issue.

I'm also developing in this style of 'Copy Sprite and modify" because running two instances of Snap! inside my web browser is really slow. Otherwise, I'd just keep the old version in a seperate tab/window as I develop the new.

I might try 'fixing' my project later by hand-editing the XML, just for fun :-).

And/or I might just export the final desired blocks individually, and import them one by one in a fresh project.

You can close this bug if you want, I'm definitely pushing Snap! hard. I doubt anyone else is going to have these problems. Though I bet your curiosity is piqued, lol!

**Comment by deader69** [comment]
BTW, I'm doing this funky "set var to lambda" thing because of that event queue bug/feature that filters out redundant events that are close in time (I was originally drawing each dot/dash via an event).

Also, events don't have parameters. That'd be a better means of intra-object communication.

**Comment by deader69** [comment]
Also also... projects take a VERY long time for me to load, yet they save very quickly.

Are you calculating a delta, and sending only the differences back to the server?

Or is there just some weird bug in my FireFox browser?
When loading, it'll prompt me 3 or 4 or 5 times with:
______________________________________________________________________________________
A script on this page may be busy, or it may have stopped responding. You can stop the script now, open the script in the debugger, or let the script continue.

Script: http://snap.berkeley.edu/snapsource/morphic.js:7061
------------------------------------------------------------------------------------------------------------------------------------------

**Comment by deader69** [comment]
It's not always that particular line, of course... another sample:

Script: http://snap.berkeley.edu/snapsource/morphic.js:2407

**Comment by jmoenig** [comment]
Ah, okay, thanks for the explanation. Well, it's certainly a bug, so at some point I guess I'll fix it :-)
When saving to the cloud I'm distinguishing between program data (scripts) and media (pictures, sounds). If you're only changing program data, only that part of the project gets uploaded (again) to the server. This should save bandwidth. Not sure why this doesn't seem to work for you. I'll investigate...

Thanks for the merciless testing!

**Comment by deader69** [comment]
The save feature you describe DOES work for me. It saves very quickly, I was just wondering how.
Skip unchanged binary resources, of course!

It's the downloads that pause my browser... but don't worry about it. I'm about to buy a new laptop, probably a ChromeBook. I use ChromeBoxes at work, and they don't prompt/dialog on download. 

Probably just my 3 year old HP Pavilion laptop, running an aging Mint Linux, with a 700G hard disk filled to capacity. :-O

**Comment by show0k** [comment]
@deader69 Even with a modern i7 cpu computer, Snap! xml parsing is very slow. No idea why. 
I feel Snap! is less slowing my computer on Chromium than Firefox but I've not benchmarked it very seriously.

**Comment by deader69** [comment]
@show0k , thanks for the info!

I've got an HP quadcore running at 800 MHz... things are slow, but I'm also a 'power user', with lots of tabs/windows open.

Other websites sometimes tax javascript heavily... sometimes my CPU fan kicks in, blasting heat out the side... I have no idea what my machine is doing, and I don't like it, but that's the era we now live in, lol!

I'm thinking of buying the latest ChromeBook with an i5 chip... it's theoretically 'weaker' from a HW standpoint, but me thinks google tuned the OS and hardware well, so it might outperform...

Who knows? I know clones are very slow, so I've learned to show/hide theme, rather than create/destroy.

And yes, I wish XML had never been invented, lol!

Is there a forum for Snap! performance issues?
